[api-major] Rework the find controller for unit testing

timvandermeij opened the pull request:
This commit series works towards a resolution for #7356 by removing the dependency on `PDFViewer` (using the event bus) and writing a basic unit test for the find controller to have a framework to build more unit tests upon.

I think #7356 can be closed once more unit tests are written after this commit series lands, in particular unit tests that exercise the different options and commands for the find controller such as phrase search, case sensitive search, whole words search, et cetera.

timvandermeij committed:
Remove the find bar's dependency on the find controller

Pull request #10100 removed the last usage of the find controller from
the find bar, so we can drop the dependency now.

timvandermeij committed:
Use the `updatetextlayermatches` event for highlighting matches on a page

This makes use of the event bus instead of requiring the PDF viewer
instance to get the page view for a page and calling `updateMatches` on
it.

timvandermeij committed:
Implement the `setDocument` method for the find controller

Now it follows the same pattern as e.g., the document properties
component, which allows us to have one instance of the find controller
and set a new document to search upon switching documents.

Moreover, this allows us to get rid of the dependency on `pdfViewer` in
order to fetch the text content for a page. This is working towards
getting rid of the `pdfViewer` dependency upon initializing the
component entirely in future commits.

Finally, we make the `reset` method private since it's not supposed to
be used from the outside anymore now that `setDocument` takes care of
this, similar to other components.

timvandermeij committed:
Use the link service for getting and setting page information

This removes the dependency on a `PDFViewer` instance from the find
controller, which makes it more similar to other components and makes it
easier to unit test with a mock link service.

Finally, we remove the search capabilities from the SVG example since it
doesn't work there because there is no separate text layer.

timvandermeij committed:
Replace callbacks for updating the UI with dispatching events on the event bus

This makes it more similar to how other components update the viewer UI
and avoids the need to have extra member variables and checks.

timvandermeij committed:
Implement a basic unit test for the find controller

This commit shows that we can now unit test the find controller and
that executing regular queries works. Note that this is only a first
step and not a complete suite of unit tests for all possible options
of the find controller.

While writing this unit test, I found two smaller issues that I
addressed directly. The first one is that in the previous find
controller refactoring I forgot to rename some occurrences of a now
private member variable. Fortunately this did not cause any bugs since
we did have a public getter and the fetched value may be changed by
reference, but it's nevertheless good to fix. The second issue is that
some entries in the `test/unit/clitests.json` file were not correct,
resulting in these tests not being executed on e.g., Travis CI.

Snuffleupagus reviewed (commented):
I *really* like the how you're making use of the `EventBus` to break apart a number of the existing component dependencies in these patches :-)

While I'm liking the overall approach, please note that I've not yet had time to really test this. However, I've left a couple of initial comments based on a cursory look at the code. Thank you for doing this refactoring!

---

Unrelated to the rest of the PR, but something that I just happened to notice: Should `this._linkService.pagesCount` be used instead at https://github.com/mozilla/pdf.js/blob/54d6c2436c17b6d0efc6fb8e36f36358eba1a9e6/web/pdf_find_controller.js#L466

Snuffleupagus commented on web/pdf_find_controller.js:
```diff
@@ -341,10 +341,10 @@ class PDFFindController {
-    if (page.textLayer) {
-      page.textLayer.updateMatches();
-    }
+    this._eventBus.dispatch('updatematches', {
```
It might not hurt to use a slightly more specific name here, so how about e.g. `updatetextlayermatches` instead.

timvandermeij replied:
Done in the new commit series.

Snuffleupagus commented:
> it should be implemented properly as tracked in issue #7468 and not partially and inconsistently as in the current state.

Please note that while I've not tried to implement it, the "only" thing preventing that issue from being solved is that there's (currently) no way of knowing, i.e. no `findbarclose` event fired, when the native browser findbar is closed (which thus prevents a consistent implementation).

timvandermeij committed:
Remove the find controller setter in `web/base_viewer.js`

With `PDFFindController` instances no longer (directly) depending on
`BaseViewer` instances, we can pass a single `findController` when
initializing a viewer, similar to other components.

timvandermeij commented on web/pdf_find_controller.js:
```diff
@@ -312,7 +333,14 @@ class PDFFindController {
 
       promise = promise.then(() => {
-        return this._pdfViewer.getPageTextContent(i).then((textContent) => {
+        const textContentPromise = this._pdfDocument.getPage(i + 1).then(
```
Done in the new commit series, together with the `numPages` comment above.

timvandermeij commented on web/app.js:
```diff
@@ -343,7 +343,7 @@ let PDFViewerApplication = {
     pdfLinkService.setHistory(this.pdfHistory);
 
     this.findController = new PDFFindController({
```
Done in the new commit series (in a separate commit) since I think it's a good idea and it's more consistent with how other components are passed in.

timvandermeij commented on web/pdf_find_controller.js:
```diff
@@ -551,19 +548,21 @@ class PDFFindController {
-    }
     const matchesCount = this._requestMatchesCount();
-    this.onUpdateResultsCount(matchesCount);
+    this._eventBus.dispatch('updatefindbarresultscount', {
```
Done in the new commit series.

timvandermeij commented on web/pdf_find_controller.js:
```diff
@@ -551,19 +548,21 @@ class PDFFindController {
-    this.onUpdateResultsCount(matchesCount);
+    this._eventBus.dispatch('updatefindbarresultscount', {
+      source: this,
+      matchesCount,
```
Done in the new commit series.

timvandermeij commented on web/pdf_find_controller.js:
```diff
@@ -551,19 +548,21 @@ class PDFFindController {
-    }
     const matchesCount = this._requestMatchesCount();
-    this.onUpdateState(state, previous, matchesCount);
+    this._eventBus.dispatch('updatefindbarstate', {
```
Done in the new commit series.

timvandermeij commented on web/pdf_find_controller.js:
```diff
@@ -551,19 +548,21 @@ class PDFFindController {
+      source: this,
+      state,
+      previous,
+      matchesCount,
```
Done in the new commit series.

timvandermeij commented on web/app.js:
```diff
@@ -1976,6 +1963,30 @@ function webViewerFindFromUrlHash(evt) {
 }
 
+function webViewerUpdateFindBarResultsCount(evt) {
+  const matchesCount = evt.matchesCount;
```
Done in the new commit series.

timvandermeij commented on web/app.js:
```diff
@@ -1976,6 +1963,30 @@ function webViewerFindFromUrlHash(evt) {
+}
+
+function webViewerUpdateFindBarState(evt) {
+  const { state, previous, matchesCount, } = evt;
```
Done in the new commit series.

timvandermeij commented on test/unit/pdf_find_controller_spec.js:
```diff
@@ -0,0 +1,93 @@
+import { getDocument } from '../../src/display/api';
+import { PDFFindController } from '../../web/pdf_find_controller';
+
+class MockLinkService {
```
Indeed, that's why I chose for a mock here. I initially used the `SimpleLinkService` class, but quickly realized that it was a bit too simple for the testing purposes.

Snuffleupagus replied:
One more tiny nit here: How about importing `SimpleLinkService` above and do `class MockLinkService extends SimpleLinkService` here, since that will first of all guarantee that all of the expected methods are present and second of all will also help "document" the `MockLinkService` w.r.t. the expected interface!?

timvandermeij commented on test/unit/pdf_find_controller_spec.js:
```diff
@@ -0,0 +1,93 @@
+    pdfFindController.executeCommand('find', { query: 'Dynamic', });
+
+    const totalPages = 14;
+    const totalMatches = 33;
```
Done in the new commit. I also changed `totalPages` to be the length of the `matchesPerPage` array instead of a hardcoded constant.

timvandermeij commented on examples/components/simpleviewer.js:
```diff
@@ -46,7 +46,7 @@ pdfLinkService.setViewer(pdfViewer);
 
 // (Optionally) enable find controller.
 var pdfFindController = new pdfjsViewer.PDFFindController({
```
D'oh! I forgot that there is no search query by default, so I totally missed this. I fixed this in the new commit series and verified that it works with a query now.

timvandermeij commented:
Thank you for the helpful review comments! I have addressed them in the new commit series.

timvandermeij commented:
/botio unittest

timvandermeij commented:
/botio-linux preview

Snuffleupagus commented on web/text_layer_builder.js:
```diff
@@ -56,6 +56,14 @@ class TextLayerBuilder {
     this.enhanceTextSelection = enhanceTextSelection;
 
+    // Listen for events about updating find matches for this page.
+    this.eventBus.on('updatetextlayermatches', (evt) => {
```
I've just realized that this PR, and my recent find highlight one, both suffer from the same *very serious* bug. The `EventBus` listeners that are bound in this way are never actually being removed, casing the `EventBus._listeners['updatetextlayermatches'] = [...]` array to grow without bounds as new `TextLayerBuilder` instances are created :-(

Simply put, just because the `this.textLayer` property on `PDFPageView` is [cleared here](https://github.com/mozilla/pdf.js/blob/54d6c2436c17b6d0efc6fb8e36f36358eba1a9e6/web/pdf_page_view.js#L272) that does *not* imply that the `TextLayerBuilder` instance actually goes away unfortunately.

Since it's less than fifteen minutes since I found this problem, I'm still not sure if there exists a *generally* viable solution here (i.e. one not requiring ugly hacks). Also, it's now becoming clear to me why the current implementation (i.e. the one in `master`) looks the way it does, given these recently discovered problems.

timvandermeij replied:
Why does the `TextLayerBuilder` instance not get deleted when it's cleared? Can't we use `this.eventBus.off` in https://github.com/mozilla/pdf.js/blob/40a6d175555316f2dbafc5bc240e52114a3e1eb7/web/text_layer_builder.js#L125 to remove the listener once rendering is being cancelled, or would that not be enough?

Snuffleupagus replied:
I did considered that, but quite frankly it's really not at all a good idea since cancelling the `renderTextLayer` task itself doesn't necessarily mean that you'll also want to subsequently "destroy" the `TextLayerBuilder` instance too; in particular consider what would happen with multiple `TextLayerBuilder.render` calls with such a change: https://github.com/mozilla/pdf.js/blob/54d6c2436c17b6d0efc6fb8e36f36358eba1a9e6/web/text_layer_builder.js#L87-L91

The best idea I'm able to come up with on short notice, is the *first* commit that's now included in PR #10100. Please note that I'm not saying that it's a great solution, but it seems to work reasonably well and having an event fire when rendering is cancelled could be generally useful too (as a complement to 'pagerendered').

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio unittest

Snuffleupagus reviewed (approved):
This looks *great*, and seem to work just fine; thank you!
